File: truevfs-kernel/truevfs-kernel-impl/src/main/scala/net/java/truevfs/kernel/impl/ArchiveModel.scala
/*
 * Copyright (C) 2005-2012 Schlichtherle IT Services.
 * All rights reserved. Use is subject to license terms.
 */
package net.java.truevfs.kernel.impl

import net.java.truevfs.kernel.spec._

/**
  * @author Christian Schlichtherle
  */
private trait ArchiveModel[E <: FsArchiveEntry]
extends FsModel with ReentrantReadWriteLockAspect {

  def driver: FsArchiveDriver[E]

  /** Composes the node path from the mountpoint of this model and the given
    * node name.
    * 
    * @param name the node name.
    */
  final def path(name: FsNodeName) = new FsNodePath(getMountPoint, name)

  def touch(options: AccessOptions)
}

File: truevfs-kernel/truevfs-kernel-impl/src/main/java/net/java/truevfs/kernel/impl/ArchiveModel.java
/*
 * Copyright (C) 2005-2020 Schlichtherle IT Services.
 * All rights reserved. Use is subject to license terms.
 */
package net.java.truevfs.kernel.impl;

import net.java.truecommons.shed.BitField;
import net.java.truevfs.kernel.spec.*;

import java.io.IOException;
import java.util.concurrent.locks.ReentrantReadWriteLock;

/**
 * @author Christian Schlichtherle
 */
abstract class ArchiveModel<E extends FsArchiveEntry>
        extends FsDecoratingModel
        implements ReentrantReadWriteLockAspect {

    final ReentrantReadWriteLock lock = new ReentrantReadWriteLock();

    private final FsArchiveDriver<E> driver;

    ArchiveModel(final FsArchiveDriver<E> driver, final FsModel model) {
        super(model);
        this.driver = driver;
    }

    final FsArchiveDriver<E> getDriver() {
        return driver;
    }

    @Override
    public ReentrantReadWriteLock getLock() {
        return lock;
    }

    /**
     * Composes the node path from the mountpoint of this model and the given node name.
     *
     * @param name the node name.
     */
    final FsNodePath path(FsNodeName name) {
        return new FsNodePath(getMountPoint(), name);
    }

    abstract void touch(BitField<FsAccessOption> options) throws IOException;
}
